Ran client system tests here, all passing except for one which looks flaky 

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-06-16--001.1592335112--mumrah--KAFKA-10123--63c1b14a4/report.html

```
====================================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.7.7
session_id:       2020-06-16--001
run time:         62 minutes 47.115 seconds
tests run:        61
passed:           60
failed:           1
ignored:          0
====================================================================================================
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=0.10.0.1
status:     PASS
run time:   1 minute 0.624 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=0.10.1.1
status:     PASS
run time:   57.225 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=0.10.2.2
status:     PASS
run time:   57.211 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=0.11.0.3
status:     PASS
run time:   59.375 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=1.0.2
status:     PASS
run time:   1 minute 1.141 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=1.1.1
status:     PASS
run time:   58.956 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=2.0.1
status:     PASS
run time:   1 minute 3.180 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=2.1.1
status:     PASS
run time:   1 minute 8.069 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=2.2.2
status:     PASS
run time:   1 minute 4.166 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=2.3.1
status:     PASS
run time:   1 minute 4.140 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=2.4.1
status:     PASS
run time:   1 minute 5.373 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_features_test.ClientCompatibilityFeaturesTest.run_compatibility_test.broker_version=dev
status:     PASS
run time:   1 minute 2.516 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=0.10.0.1
status:     PASS
run time:   2 minutes 21.015 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=0.10.1.1
status:     PASS
run time:   2 minutes 25.114 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=0.10.2.2
status:     PASS
run time:   2 minutes 20.943 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=0.11.0.3
status:     PASS
run time:   2 minutes 27.628 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=1.0.2
status:     PASS
run time:   2 minutes 23.814 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=1.1.1
status:     PASS
run time:   2 minutes 25.035 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=2.0.1
status:     PASS
run time:   2 minutes 26.260 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=2.1.1
status:     PASS
run time:   2 minutes 30.776 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=2.2.2
status:     PASS
run time:   2 minutes 29.093 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=2.3.1
status:     PASS
run time:   2 minutes 34.028 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=2.4.1
status:     PASS
run time:   2 minutes 32.940 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.client_compatibility_produce_consume_test.ClientCompatibilityProduceConsumeTest.test_produce_consume.broker_version=dev
status:     PASS
run time:   2 minutes 28.548 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.pluggable_test.PluggableConsumerTest.test_start_stop
status:     PASS
run time:   20.868 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.AssignmentValidationTest.test_valid_assignment.assignment_strategy=org.apache.kafka.clients.consumer.RangeAssignor
status:     PASS
run time:   55.992 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.AssignmentValidationTest.test_valid_assignment.assignment_strategy=org.apache.kafka.clients.consumer.RoundRobinAssignor
status:     PASS
run time:   55.605 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.AssignmentValidationTest.test_valid_assignment.assignment_strategy=org.apache.kafka.clients.consumer.StickyAssignor
status:     PASS
run time:   56.918 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.message_format_change_test.MessageFormatChangeTest.test_compatibility.consumer_version=0.9.0.1.producer_version=0.9.0.1
status:     PASS
run time:   3 minutes 47.147 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.message_format_change_test.MessageFormatChangeTest.test_compatibility.consumer_version=0.10.2.2.producer_version=0.10.2.2
status:     PASS
run time:   4 minutes 42.586 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=all
status:     PASS
run time:   57.221 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.consumer_num=2.quota_type=client-id
status:     PASS
run time:   3 minutes 9.596 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=stable
status:     PASS
run time:   1 minute 7.739 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=all
status:     PASS
run time:   1 minute 4.560 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.compression_test.CompressionTest.test_compressed_topic.compression_types=.snappy.gzip.lz4.zstd.none
status:     PASS
run time:   2 minutes 7.014 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.message_format_change_test.MessageFormatChangeTest.test_compatibility.consumer_version=dev.producer_version=dev
status:     PASS
run time:   4 minutes 22.235 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=stable
status:     FAIL
run time:   1 minute 9.961 seconds


    
Traceback (most recent call last):
  File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/lib/python2.7/site-packages/ducktape-0.7.7-py2.7.egg/ducktape/tests/runner_client.py", line 132, in run
    data = self.run_test()
  File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/lib/python2.7/site-packages/ducktape-0.7.7-py2.7.egg/ducktape/tests/runner_client.py", line 189, in run_test
    return self.test_context.function(self.test)
  File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/venv/lib/python2.7/site-packages/ducktape-0.7.7-py2.7.egg/ducktape/mark/_mark.py", line 428, in wrapper
    return functools.partial(f, *args, **kwargs)(*w_args, **w_kwargs)
  File "/home/jenkins/workspace/system-test-kafka-branch-builder/kafka/tests/kafkatest/tests/client/consumer_test.py", line 257, in test_fencing_static_consumer
    assert len(consumer.dead_nodes()) == num_conflict_consumers
AssertionError

----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_broker_failure.clean_shutdown=False.enable_autocommit=False
status:     PASS
run time:   1 minute 1.026 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_broker_failure.clean_shutdown=False.enable_autocommit=True
status:     PASS
run time:   1 minute 0.342 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_rolling_upgrade_test.ConsumerRollingUpgradeTest.rolling_update_test
status:     PASS
run time:   40.060 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_broker_failure.clean_shutdown=True.enable_autocommit=False
status:     PASS
run time:   55.844 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_broker_failure.clean_shutdown=True.enable_autocommit=True
status:     PASS
run time:   56.067 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.old_broker_throttling_behavior=True.quota_type=client-id
status:     PASS
run time:   3 minutes 13.626 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_consumer_bounce.clean_shutdown=True.bounce_mode=all
status:     PASS
run time:   1 minute 25.122 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_broker_rolling_bounce
status:     PASS
run time:   2 minutes 49.878 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_consumer_bounce.clean_shutdown=True.bounce_mode=rolling
status:     PASS
run time:   2 minutes 15.369 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_consumer_failure.clean_shutdown=True.enable_autocommit=False
status:     PASS
run time:   1 minute 0.818 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_consumer_failure.clean_shutdown=True.enable_autocommit=True
status:     PASS
run time:   1 minute 0.885 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_group_consumption
status:     PASS
run time:   1 minute 5.476 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_static_consumer_bounce.static_membership=True.clean_shutdown=True.bounce_mode=all.num_bounces=5
status:     PASS
run time:   1 minute 20.141 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_static_consumer_bounce.static_membership=False.clean_shutdown=True.bounce_mode=all.num_bounces=5
status:     PASS
run time:   1 minute 34.796 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_static_consumer_bounce.static_membership=False.clean_shutdown=True.bounce_mode=rolling.num_bounces=5
status:     PASS
run time:   1 minute 41.124 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_static_consumer_bounce.static_membership=True.clean_shutdown=True.bounce_mode=rolling.num_bounces=5
status:     PASS
run time:   1 minute 32.181 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.truncation_test.TruncationTest.test_offset_truncate
status:     PASS
run time:   1 minute 43.953 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=.user.client-id.override_quota=True
status:     PASS
run time:   3 minutes 18.792 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=client-id.old_client_throttling_behavior=True
status:     PASS
run time:   3 minutes 28.109 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=.user.client-id.override_quota=False
status:     PASS
run time:   4 minutes 15.238 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=client-id.override_quota=True
status:     PASS
run time:   3 minutes 20.559 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=client-id.override_quota=False
status:     PASS
run time:   4 minutes 16.169 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=user.override_quota=False
status:     PASS
run time:   4 minutes 10.209 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.quota_test.QuotaTest.test_quota.quota_type=user.override_quota=True
status:     PASS
run time:   3 minutes 13.171 seconds
----------------------------------------------------------------------------------------------------
```

Re-ran the flaky test here: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-06-16--001.1592340267--mumrah--KAFKA-10123--e12d62f27/report.html

```
====================================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.7.7
session_id:       2020-06-16--001
run time:         6 minutes 4.729 seconds
tests run:        16
passed:           16
failed:           0
ignored:          0
====================================================================================================
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=all
status:     PASS
run time:   56.582 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=all
status:     PASS
run time:   1 minute 2.261 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=stable
status:     PASS
run time:   1 minute 4.217 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=all
status:     PASS
run time:   55.329 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=stable
status:     PASS
run time:   1 minute 10.086 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=stable
status:     PASS
run time:   1 minute 10.305 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=all
status:     PASS
run time:   57.946 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=stable
status:     PASS
run time:   1 minute 6.535 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=all
status:     PASS
run time:   1 minute 3.113 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=stable
status:     PASS
run time:   1 minute 1.881 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=all
status:     PASS
run time:   1 minute 2.480 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=stable
status:     PASS
run time:   1 minute 8.963 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=all
status:     PASS
run time:   59.126 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=1.fencing_stage=stable
status:     PASS
run time:   1 minute 4.819 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=all
status:     PASS
run time:   1 minute 3.293 seconds
----------------------------------------------------------------------------------------------------
test_id:    kafkatest.tests.client.consumer_test.OffsetValidationTest.test_fencing_static_consumer.num_conflict_consumers=2.fencing_stage=stable
status:     PASS
run time:   1 minute 7.971 seconds
----------------------------------------------------------------------------------------------------
```

--- a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
+++ b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
@@ -783,9 +783,9 @@ public final class ConsumerCoordinator extends AbstractCoordinator {
      * @return true iff the operation completed within the timeout
      */
     public boolean refreshCommittedOffsetsIfNeeded(Timer timer) {
-        final Set<TopicPartition> missingFetchPositions = subscriptions.initializingPartitions();
+        final Set<TopicPartition> initializingPartitions = subscriptions.initializingPartitions();
 
-        final Map<TopicPartition, OffsetAndMetadata> offsets = fetchCommittedOffsets(missingFetchPositions, timer);
+        final Map<TopicPartition, OffsetAndMetadata> offsets = fetchCommittedOffsets(initializingPartitions, timer);
         if (offsets == null) return false;
 
         for (final Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {

--- a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
+++ b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
@@ -1292,7 +1292,7 @@ public class Fetcher<K, V> implements Closeable {
                 if (!clearedReplicaId.isPresent()) {
                     // If there's no preferred replica to clear, we're fetching from the leader so handle this error normally
                     FetchPosition position = subscriptions.position(tp);
-                    if (position != null && fetchOffset != position.offset) {
+                    if (position == null || fetchOffset != position.offset) {
                         log.debug("Discarding stale fetch response for partition {} since the fetched offset {} " +
                                 "does not match the current offset {}", tp, fetchOffset, position);
                     } else {

--- a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java
+++ b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java
@@ -792,7 +792,7 @@ public class SubscriptionState {
          * also will update the position with the current leader and epoch.
          *
          * @param currentLeaderAndEpoch leader and epoch to compare the offset with
-         * @return
+         * @return true if the position is now awaiting validation
          */
         private boolean maybeValidatePosition(Metadata.LeaderAndEpoch currentLeaderAndEpoch) {
             if (this.fetchState.equals(FetchStates.AWAIT_RESET)) {
@@ -813,14 +813,11 @@ public class SubscriptionState {
 
         /**
          * For older versions of the API, we cannot perform offset validation so we simply transition directly to FETCHING
-         *
-         * @param currentLeaderAndEpoch
          */
         private void updatePositionLeaderNoValidation(Metadata.LeaderAndEpoch currentLeaderAndEpoch) {
             if (position != null) {
-                FetchPosition newPosition = new FetchPosition(position.offset, position.offsetEpoch, currentLeaderAndEpoch);
                 transitionState(FetchStates.FETCHING, () -> {
-                    this.position = newPosition;
+                    this.position = new FetchPosition(position.offset, position.offsetEpoch, currentLeaderAndEpoch);
                     this.nextRetryTimeMs = null;
                 });
             }
@@ -957,7 +954,7 @@ public class SubscriptionState {
         Collection<FetchState> validTransitions();
 
         /**
-         * Test if this state has a position
+         * Test if this state requires a position to be set
          */
         boolean requiresPosition();
 

--- a/clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java
+++ b/clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java
@@ -1190,7 +1190,7 @@ public class FetcherTest {
         assertEquals(0, fetcher.fetchedRecords().size());
         assertTrue(subscriptions.isOffsetResetNeeded(tp0));
         assertNull(subscriptions.validPosition(tp0));
-        assertNotNull(subscriptions.position(tp0));
+        assertNull(subscriptions.position(tp0));
     }
 
     @Test
@@ -3854,8 +3854,6 @@ public class FetcherTest {
             return partitionData.currentLeaderEpoch.equals(Optional.of(epochOne)) && partitionData.leaderEpoch == epochOne;
         }, new OffsetsForLeaderEpochResponse(singletonMap(tp0, epochEndOffset)));
         consumerClient.poll(time.timer(Duration.ZERO));
-
-        assertEquals(initialOffset, subscriptions.position(tp0).offset);
 
         if (offsetResetStrategy == OffsetResetStrategy.NONE) {
             OffsetOutOfRangeException thrown =

--- a/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
+++ b/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
@@ -42,7 +42,6 @@ import java.util.regex.Pattern;
 import static java.util.Collections.singleton;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 

--- a/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
+++ b/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
@@ -43,6 +43,7 @@ import static java.util.Collections.singleton;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
 public class SubscriptionStateTest {
@@ -210,7 +211,7 @@ public class SubscriptionStateTest {
         state.requestOffsetReset(tp0);
         assertFalse(state.isFetchable(tp0));
         assertTrue(state.isOffsetResetNeeded(tp0));
-        assertNotNull(state.position(tp0));
+        assertNull(state.position(tp0));
 
         // seek should clear the reset and make the partition fetchable
         state.seek(tp0, 0);
@@ -600,7 +601,7 @@ public class SubscriptionStateTest {
         assertEquals(Optional.empty(), divergentOffsetMetadataOpt);
         assertFalse(state.awaitingValidation(tp0));
         assertTrue(state.isOffsetResetNeeded(tp0));
-        assertEquals(initialPosition, state.position(tp0));
+        assertNull(state.position(tp0));
     }
 
     @Test

--- a/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
+++ b/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java
@@ -704,6 +704,22 @@ public class SubscriptionStateTest {
         assertTrue(state.hasValidPosition(tp0));
         assertFalse(state.awaitingValidation(tp0));
         assertFalse(state.isOffsetResetNeeded(tp0));
+
+        // Reset again, and complete it with a seek that would normally require validation
+        state.requestOffsetReset(tp0, OffsetResetStrategy.EARLIEST);
+        state.seekUnvalidated(tp0, new SubscriptionState.FetchPosition(10L, Optional.of(10), new Metadata.LeaderAndEpoch(
+                Optional.of(broker1), Optional.of(2))));
+        // We are now in AWAIT_VALIDATION
+        assertFalse(state.hasValidPosition(tp0));
+        assertTrue(state.awaitingValidation(tp0));
+        assertFalse(state.isOffsetResetNeeded(tp0));
+
+        // Now ensure next call to validate clears the validation state
+        assertFalse(state.maybeValidatePositionForCurrentLeader(oldApis, tp0, new Metadata.LeaderAndEpoch(
+                Optional.of(broker1), Optional.of(2))));
+        assertTrue(state.hasValidPosition(tp0));
+        assertFalse(state.awaitingValidation(tp0));
+        assertFalse(state.isOffsetResetNeeded(tp0));
     }
 
 }

